File: packages/core/src/shell/builtins/pwd.ts
/**
 * pwd - print working directory builtin command
 */

import { BuiltinCommand, ShellEnvironment, ShellOptions } from '../types.js';

export class PwdCommand implements BuiltinCommand {
  name = 'pwd';
  description = 'Print the current working directory';

  async execute(args: string[], env: ShellEnvironment, options: ShellOptions): Promise<number> {
    // Handle -L (logical) and -P (physical) options
    const useLogical = !args.includes('-P');
    
    try {
      if (useLogical) {
        // Use the PWD environment variable (logical path)
        console.log(env.getVariable('PWD') || env.cwd);
      } else {
        // Use the actual physical path
        console.log(process.cwd());
      }
      
      return 0;
    } catch (error) {
      console.error(`pwd: ${error instanceof Error ? error.message : String(error)}`);
      return 1;
    }
  }

  help(): string {
    return `pwd [-L | -P]
Print the absolute pathname of the current working directory.

Options:
  -L    Print the logical current directory (default)
  -P    Print the physical current directory (all symbolic links resolved)`;
  }
}

File: shell/src/shell/builtins/pwd.rs
//! pwd - print working directory builtin command

use crate::shell::types::*;

pub struct PwdCommand;

#[async_trait::async_trait]
impl BuiltinCommand for PwdCommand {
    fn name(&self) -> &str {
        "pwd"
    }

    fn description(&self) -> &str {
        "Print the current working directory"
    }

    async fn execute(
        &self,
        _args: &[String],
        env: &mut ShellEnvironment,
        _options: &ShellOptions,
    ) -> Result<i32, ShellError> {
        println!("{}", env.cwd);
        Ok(0)
    }

    fn help(&self) -> String {
        "pwd
Print the absolute pathname of the current working directory.".to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_pwd() {
        let mut env = ShellEnvironment::new();
        env.cwd = "/test/directory".to_string();
        
        let pwd = PwdCommand;
        let result = pwd.execute(&[], &mut env, &ShellOptions::default()).await;
        
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), 0);
    }
}